pkg/handler/collector/s3/s3.go

```diff
@@ -0,0 +1,176 @@
+//
+// Copyright 2023 The GUAC Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package s3
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/guacsec/guac/pkg/handler/collector/s3/bucket"
+	"github.com/guacsec/guac/pkg/handler/collector/s3/messaging"
+	"github.com/guacsec/guac/pkg/handler/processor"
+	"github.com/guacsec/guac/pkg/logging"
+)
+
+const (
+	S3CollectorType = "S3CollectorType"
+)
+
+type S3Collector struct {
+	config S3CollectorConfig
+}
+
+type S3CollectorConfig struct {
+	MessageProvider     string
+	MessageProviderHost string
+	MessageProviderPort string
+	S3Host              string
+	S3Port              string
+	Region              string // optional (defaults to us-east-1, assumes same region for s3 and sqs)
+	Queues              string
+	MpBuilder           messaging.MessageProviderBuilder // optional
+	BucketBuilder       bucket.BuildBucket               // optional
+	SigChan             chan os.Signal                   // optional
+}
+
+func NewS3Collector(cfg S3CollectorConfig) (*S3Collector, error) {
+	s3collector := &S3Collector{
+		config: cfg,
+	}
+	return s3collector, nil
+}
+
+func (s S3Collector) RetrieveArtifacts(ctx context.Context, docChannel chan<- *processor.Document) error {
+	logger := logging.FromContext(ctx)
+	queues := strings.Split(s.config.Queues, ",")
+	downloader := getDownloader(s)
+	sigChan := s.config.SigChan
+
+	var wg sync.WaitGroup
+	for _, queue := range queues {
+		wg.Add(1)
+
+		go func(queue string) {
+			mp, err := getMessageProvider(s, queue)
+			if err != nil {
+				logger.Errorf("error getting message provider for queue %v: %v", queue, err)
+				return
+			}
+
+			for {
+				select {
+				case <-sigChan:
+					logger.Infof("Shutting down collector for queue %s...\n", queue)
+					break
+				default:
+					m, err := mp.ReceiveMessage(ctx)
+					if err != nil {
+						logger.Infof("error while receiving message, skipping: %v\n", err)
+						continue
+					}
+
+					if e, er := m.GetEvent(); e != messaging.PUT {
+						if er != nil {
+							logger.Debugf("skipping message: %v\n", er)
+						}
+						continue
+					}
+					bucketName, err := m.GetBucket()
+					if err != nil {
+						logger.Errorf("skipping message: %v\n", err)
+						continue
+					}
+					item, err := m.GetItem()
+					if err != nil {
+						logger.Errorf("skipping message: %v\n", err)
```

Comment:
I think we need a `continue` to prevent processing with potentially invalid data.

Reply:
Thank you, good catch